api/rtx/src/uvisor_allocator.c

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) 2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "uvisor_allocator.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include "uvisor-lib/uvisor-lib.h"
+#include "rt_TypeDef.h"
+#include "rt_Memory.h"
+#include <string.h>
+
+#define DPRINTF(...) {};
+/* #define DPRINTF(fmt, ...) printf(fmt, ## __VA_ARGS__) */
+
+/* Declare this variable here, so the tier-2 allocator _always_ uses the
+ * page size that the tier-1 allocator expects! */
+const uint32_t __uvisor_page_size = UVISOR_PAGE_SIZE;
+
+/* Internal structure currently only contains the page table */
+typedef struct {
+    UvisorPageTable table;
+} UvisorAllocatorInternal;
+
+static inline UvisorPageTable * table(UvisorAllocator allocator) {
+    return &(((UvisorAllocatorInternal *) allocator)->table);
+}
+
+UvisorAllocator uvisor_allocator_create_with_pool(
+    void * mem,
+    size_t bytes)
+{
+    UvisorAllocatorInternal * allocator = mem;
+    /* Signal that this is non-page allocated memory */
+    allocator->table.page_size = bytes;
+    allocator->table.page_count = 0;
+    /* The MEMP structure must be placed AFTER table.page_origins[0] !!! */
+    size_t offset = offsetof(UvisorAllocatorInternal, table.page_origins) + sizeof(void *);
+    /* Create MEMP structure inside the memory */
+    if (rt_init_mem(mem + offset, bytes - offset)) {
+        /* Abort if failed */
+        DPRINTF("uvisor_allocator_create_with_pool: MEMP allocator creation failed\n\n");
+        return NULL;
+    }
+    /* Remember the MEMP pointer though */
+    allocator->table.page_origins[0] = mem + offset;
+    DPRINTF("uvisor_allocator_create_with_pool: Created MEMP allocator %p with offset %d\n\n", mem + offset, offset);
+    return allocator;
+}
+
+UvisorAllocator uvisor_allocator_create_with_pages(
+    size_t size,
+    size_t maximum_malloc_size)
+{
+    /* Additional overhead for each block */
+    const size_t block_overhead = 2 * sizeof(MEMP);
+    size_t page_count = 0;
+
+    /* Calcutate the required total size considering page size and overhead */
+    for (int request = size; request > 0; request -= UVISOR_PAGE_SIZE, page_count++)
```

Comment:
Yes, this is a leftover from a previous memory algorithm that didn't have a fixed overhead per page.
I believe this code computes the same thing:
```c
    /* Calculate the integer part of required the page count. */
    size_t page_count = size / page_size_with_overhead;
    /* Add another page if the remainder is not zero. */
    if (size - page_count * page_size_with_overhead) {
        page_count++;
    }
```